File: swixUseCases/swix-iOS/swix-iOS/swix/numbers.swift
//
//  constants.swift
//  swix
//
//  Created by Scott Sievert on 7/9/14.
//  Copyright (c) 2014 com.scott. All rights reserved.
//

import Foundation
import Accelerate

// should point to the swift folder
let S2_PREFIX = "\(NSHomeDirectory())/Developer/swix/swix/swix/swix/"
let PYTHON_PATH = "~/anaconda/bin/ipython"

let pi = 3.1415926535897932384626433832795028841971693993751058
let π = pi
let phi = (1 + sqrt(5.double))/2
let φ = phi
let e = exp(1.double)
let euler = 0.57721566490153286060651209008240243104215933593992
let inf = 1.0 / 0.0
let DOUBLE_EPSILON = DBL_EPSILON // smallest possible difference
let FLOAT_EPSILON = FLT_EPSILON

// use 3.double or 3.14.int or N.int
extension Int{
    var int:Int {return Int(self)}
    var cint:CInt {return CInt(self)}
    var float:Float {return Float(self)}
    var double:Double {return Double(self)}
}
extension Double{
    var int:Int {return Int(self)}
    var float:Float {return Float(self)}
    var double:Double {return Double(self)}
    var cdouble:CDouble {return CDouble(self)}
}
extension CInt{
    var int:Int {return Int(self)}
    var float:Float {return Float(self)}
    var double:Double {return Double(self)}
}
extension Float{
    var int:Int {return Int(self)}
    var cfloat:CFloat {return CFloat(self)}
    var float:Float {return Float(self)}
    var double:Double {return Double(self)}

}
extension String {
    var floatValue: Float {
        return (self as NSString).floatValue
    }
    var doubleValue: Double {
        return (self as NSString).doubleValue
    }
    var nsstring:NSString {return NSString(string:self)}
}
func close(x: Double, y: Double)->Bool{
    return abs(x-y)<1e-9 ? true : false}

// damn integer division causes headaches
infix operator  / {associativity none precedence 140}
func / (lhs: Int, rhs: Int) -> Double{
    return lhs.double / rhs.double}

// a quick hack to get what I want
func isNumber(x: Double)   ->Bool{return true}
func isNumber(x: Float)    ->Bool{return true}
func isNumber(x: Int)      ->Bool{return true}
func isNumber(x: CInt)     ->Bool{return true}
func isNumber(x: ndarray)   ->Bool{return false}
func isNumber(x: matrix) ->Bool{return false}
func isNumber(x: AnyObject)->Bool{return false}

File: docs/autodoc/numbers.py
# -*- coding: utf-8 -*-
import numpy as np
from math import sqrt

#: The path that should point to the swix directory. The file imshow.py, numbers.swix would be found by appending to this path.
S2_PREFIX = "\(NSHomeDirectory())/Developer/swix/swix/swix/swix/"

#: Where is Python found? Used for imshow and savefig (testing on MacOSX only!). Assumes `pylab <http://wiki.scipy.org/PyLab>`_ is installed.
PYTHON_PATH = "~/anaconda/bin/ipython"

#: The `golden ratio <https://en.wikipedia.org/wiki/The_golden_ratio>`_
#: Also callable with φ
phi = (1 + sqrt(5))/2

#: The classic `circle constant <https://en.wikipedia.org/wiki/Pi>`_. Also called π
pi = np.pi

#: The more intuitive circle constant. Also τ
tau = 2 * np.pi

#: `Euler's number <https://en.wikipedia.org/wiki/E_(mathematical_constant)>`_
e = np.e

#: `Euler's constant <https://en.wikipedia.org/wiki/Euler–Mascheroni_constant>`_
euler = 0.57721566490153286060651209008240243104215933593992

#: Infinity.
inf = np.inf

#: The smallest possible difference between two doubles. See `machine epsilon
#: <http://en.wikipedia.org/wiki/Machine_epsilon#Approximation_using_C>`_ for more
#: detail.
DOUBLE_EPSILON = "smaller"

#: The smallest possible difference between two floats. See `machine epsilon 
#: <http://en.wikipedia.org/wiki/Machine_epsilon#Approximation_using_C>`_ for more
#: detail.
FLOAT_EPSILON = "small"

def close(x, y):
    """
    :param x: Double. An input value.
    :param y: Double. Another value.

    :rtype: Bool. True only if the two values are close numerically.
    Specifically, true only if :math:`|x-y|<10^{-9}`.
    
    >>> assert(close(0, 1e-10) == true)
    """

def convience_elements():
    """
    The built in class numeric classes has convience elements.

    * {Float|Double|Int|CInt}.int. Returns the integer value
    * {Float|Double|Int|CInt}.float. The floating point value
    * {Float|Double|Int|CInt}.double. The double value.
    * String.floatValue, String.doubleValue. Converts the string to a numeric value by interpetting a NSString.
    * String.nsstring. For interfacing with C.

    Additionally, Int has an element cint, Double has element cdouble, Float
    has element cfloat.

    >>> assert(3.14.int == 3)
    >>> assert("3.14".floatValue == 3.14)
    """


def operator_slash(lhs, rhs):
    """
    The operator ``/`` that represents division. Converts integer division to floating point division.

    :param lhs: Int.
    :param rhs: Int.
    :rtype: Double = lhs / rhs
    """

def isNumber(x):
    """
    :param x: AnyObject.
    :rtype: Bool. True only if x is a number (a float, double, cint, int). Returns false for ndarray and matrix.
    """


